### DISCUSSION TITLE
Clarification on what version of JSON schema is being used

### DISCUSSION THREAD

**Discussion by G-Rath** [eslint/eslint#17573] [Q&A]
The documentation says [here](https://eslint.org/docs/latest/extend/custom-rules#options-schemas) that currently draft-04 of the JSON schema is what's being used and that there are no plans to change that due to ecosystem compatibility concerns.

However I'm finding that at least both ESLint v7 and v8 support at least some of draft-06 - using `eslint-plugin-local` with the following:

```
// .eslintplugin/customRule.js
module.exports = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Description of the rule'
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        propertyNames: { enum: ['one', 'two', 'three'] } // https://json-schema.org/understanding-json-schema/reference/object.html#property-names
      },
      {
        type: 'array',
        contains: { type: 'number' } // https://json-schema.org/understanding-json-schema/reference/array.html#contains
      },
      {
        const: 'value' // https://json-schema.org/understanding-json-schema/reference/generic.html#constant-values
      }
    ]
  },
  create: function (context) {
    return {
      // callback functions
    };
  }
};

// .eslintplugin/index.js
exports.rules = {
  'customRule': require('./customRule.js')
}

// .eslintrc.js
module.exports = {
  plugins: ['local'],
  rules: {
    'local/customRule': /* ... */
  }
}
```

Running `npx eslint .` with different options in place of the `/* ... */ ` in `.eslintrc.js`:

```
// with ['error', { five: '' }]
Oops! Something went wrong! :(

ESLint: 7.32.0

Error: .eslintrc.js:
        Configuration for rule "local/customRule" is invalid:
        Value "five" should be equal to one of the allowed values.
        Value {"five":""} property name 'five' is invalid.

// with ['error', { one: '' }, ['not-a-number']]
Oops! Something went wrong! :(

ESLint: 7.32.0

Error: .eslintrc.js:
        Configuration for rule "local/customRule" is invalid:
        Value ["not-a-number"] should contain a valid item.

// with ['error', { one: '' }, [1], 'nope']
Oops! Something went wrong! :(

ESLint: 7.32.0

Error: .eslintrc.js:
        Configuration for rule "local/customRule" is invalid:
        Value "nope" should be equal to constant.
```

I came across this when looking into updating eslint-plugin-jest to use v6 of `@typescript-eslint/utils` which removed the types for JSONSchema7 and then started erroring on `propertyNames` [here](https://github.com/jest-community/eslint-plugin-jest/blob/main/src/rules/valid-title.ts#L149-L162) which we've been using for quite a while and works fine.

cc @bradzacher @JoshuaKGoldberg

**Comment by bradzacher** [comment]
> to use v6 of @typescript-eslint/utils which removed the types for JSONSchema7

Note we didn't remove anything! The old types (the `JSONSchema4` type from `@types/json-schema`) was defined with `[key: string]: any` - hence it allowed any old garbage that may or may not work.

In v6 we just removed this and tightened the types to only allow defined properties.

We did this because we found that some of our rules had used features that weren't supported by ESLint and thus we had rules that accepted any value without validation. Tightening the types helped us find and fix things.

**Reply by G-Rath** [reply]
[You removed the JSONSchema7 types](https://npmfs.com/package/@typescript-eslint/utils/6.0.0/dist/json-schema.d.ts) that were [previously being exported](https://npmfs.com/package/@typescript-eslint/utils/5.62.0/dist/json-schema.d.ts) from `json-schema`.

fwiw I totally support the type tightening :)

**Reply by bradzacher** [reply]
Oh yeah but all the util types we had only used the v4 types - so if you were using our infra you were using v4.

**Comment by mdjermanovic** [comment]
Indeed it seems that draft-06 is supported. This will require further investigation.

**Comment by G-Rath** [comment]
I've tried digging into it further and while my original reproduction continues to work on all versions of ESLint beyond v7 (including the v9 beta), when I extract out the validation code I've found into a dedicated script it seems to only care about the violation of `const` (which was still only introduced in schema 6):

```
// from https://github.com/eslint/eslint/blob/main/lib/shared/ajv.js
const Ajv = require('ajv'),
  metaSchema = require('ajv/lib/refs/json-schema-draft-04.json');

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

const ajvImport = (additionalOptions = {}) => {
  const ajv = new Ajv({
    meta: false,
    useDefaults: true,
    validateSchema: false,
    missingRefs: 'ignore',
    verbose: true,
    schemaId: 'auto',
    ...additionalOptions,
  });

  ajv.addMetaSchema(metaSchema);

  ajv._opts.defaultMeta = metaSchema.id;

  return ajv;
};

// from https://github.com/eslint/eslint/blob/main/lib/config/rule-validator.js#L13
const ajv = ajvImport();

// from https://github.com/eslint/eslint/blob/main/lib/config/rule-validator.js#L156
const validateRule = ajv.compile({
  type: 'array',
  items: [
    { type: 'object', propertyNames: [Object] },
    { type: 'array', contains: [Object], maxContains: 2 },
    { const: 'value' },
  ],
  minItems: 0,
  maxItems: 3,
});

// from https://github.com/eslint/eslint/blob/main/lib/config/rule-validator.js#L167
const validate = ruleOptions => {
  return [validateRule(ruleOptions.slice(1)), validateRule.errors];
};

// true null
console.log(...validate(['error', { five: '' }]));

// true null
console.log(...validate(['error', { one: '' }, ['not-a-number']]));

// false [
//   {
//     keyword: 'const',
//     dataPath: '[2]',
//     schemaPath: '#/items/2/const',
//     params: { allowedValue: 'value' },
//     message: 'should be equal to constant',
//     schema: 'value',
//     parentSchema: { const: 'value' },
//     data: 'nope'
//   }
// ]
console.log(...validate(['error', { one: '' }, [1], 'nope']));
```

Note I've referenced code in `rule-validator.js` which seems to only be used for flat config and that in some versions of ESLint validation is actually coming from `ConfigValidator` in `@eslint/eslintrc` but from what I can see that looks like just a copy of `rule-validator.js`.

@epoberezkin could you maybe shed any light on this?

**Comment by epoberezkin** [comment]
We did try to migrate ESLint couple years ago indeed, but the concern was that many plugins would use the old version of JSON schema and it would be hard at the time. Most likely can be revisited now. It was and will probably still be some work, from my memory of trying to do it at the time...
